core/unix/signal.c

```diff
@@ -2676,12 +2675,16 @@ thread_set_self_context(void *cxt)
     memset(&frame, 0, sizeof(frame));
 #ifdef LINUX
 #    ifdef X86
+    dcontext_t *dcontext = get_thread_private_dcontext();
     byte *xstate = get_and_initialize_xstate_buffer(dcontext);
     frame.uc.uc_mcontext.fpstate = &((kernel_xstate_t *)xstate)->fpstate;
 #    endif /* X86 */
     frame.uc.uc_mcontext = *sc;
 #endif
+    IF_ARM(ASSERT_NOT_TESTED());
+#if defined(LINUX) && defined(X86)
```

Comment:
The prior code called this for non-Linux too (e.g., MacOS) so this does not seem like a safe change.

Reply:
Non-Linux seems to skip `save_fpstate` at other spots too.
https://github.com/DynamoRIO/dynamorio/blob/420adb01d466e134037b1db6ed7023749c3f8034/core/unix/signal.c#L5575
Do we need a fix there as well?

Reply:
Huh, that's weird.  I would expect Mac on x86 to need the same thing, yes, for the x87 state which is not in the mcontext and so can't be updated from there.  Wondering if changing that will hit the assert on the mac tests though.

Reply:
macOS doesn't implement `save_fpstate`
https://github.com/DynamoRIO/dynamorio/blob/a5ab87c9c24f176a474dc1b7d3cb4cc4fe4d8ed4/core/unix/signal_macos.c#L145

I modified the preprocessor guard to allow `save_fpstate` for macOS builds, but macOS still passes on Travis: https://travis-ci.com/github/DynamoRIO/dynamorio/jobs/425890837.

Reply:
In a bunch of spots in `signal.c`, I see that fpstate handling is done only for `X86 && LINUX` (not including macOS). Do we need to re-evaluate this, perhaps as part of a separate issue?

Reply:
The signal frame extended state layout is different: maybe you're seeing code specific to how Linux lays it out.  The general step of updating the fpstate on a delayed signal to have the new delivery spot's state is what is common and that should only be in a couple of places calling save_fpstate.   I'm guessing you're looking at different code that is not related to that step.

Reply:
Oh okay. 

The other `save_fpstate` instance[1] in `signal.c` is invoked during `receive_pending_signal`, so we do expect that to be done for macOS too right? 

About this instance of `save_fpstate` not being hit in macOS tests: actually the reset tests don't seem to be enabled for macOS. Do we want them to be?

[1]: https://github.com/DynamoRIO/dynamorio/blob/420adb01d466e134037b1db6ed7023749c3f8034/core/unix/signal.c#L5575

Reply:
Discussed offline. 

Added an XXX comment for now for the other `save_fpstate`. Will investigate in a separate PR if we need any changes there.